src/types.ts

```diff
@@ -78,16 +78,23 @@ export type GetDataType<
   K extends DataTypeKey,
   Fallback extends JsonValue = undefined,
 > = K extends keyof ProtocolMap
-  ? ProtocolMap[K] extends ProtocolWithReturn<infer Data, any>
-    ? Data
-    : ProtocolMap[K]
-  : Fallback
+  ? ProtocolMap[K] extends (...args: infer Args) => any
+    ? Args['length'] extends 0
+      ? undefined
+      : Args[0]
+    : ProtocolMap[K] extends ProtocolWithReturn<infer Data, any>
+      ? Data
+      : ProtocolMap[K]
+  : Fallback;
+
 
 export type GetReturnType<
-  K extends DataTypeKey,
-  Fallback extends JsonValue = undefined,
+  K extends DataTypeKey, 
+  Fallback extends JsonValue = undefined
 > = K extends keyof ProtocolMap
-  ? ProtocolMap[K] extends ProtocolWithReturn<any, infer Return>
-    ? Return
-    : void
-  : Fallback
+  ? ProtocolMap[K] extends (...args: any[]) => infer R
+    ? R
+    : ProtocolMap[K] extends ProtocolWithReturn<any, infer Return>
+      ? Return
+      : void
+  : Fallback;
```

Comment:
Empty line and end of file
```suggestion
  : Fallback;

```